src/Adapter/CartRule/Repository/CartRuleRepository.php

```diff
@@ -87,6 +90,162 @@ public function get(CartRuleId $cartRuleId): CartRule
+                    ])
+                    ->setParameter('productRuleGroupId', $productRuleGroupId)
+                    ->setParameter('type', $restrictionRule->type)
+                    ->execute()
+                ;
+
+                $productRuleId = $this->connection->lastInsertId();
+                $productRuleValues = [];
+                $checkedIds = [];
+                foreach ($restrictionRule->ids as $id) {
+                    if (in_array($id, $checkedIds, true)) {
+                        // skip in case there are duplicates
+                        continue;
+                    }
+                    $productRuleValues[] = sprintf('(%d, %d)', $productRuleId, $id);
+                    $checkedIds[] = $id;
+                }
+
+                $this->connection->prepare('
+                    INSERT INTO ' . $this->dbPrefix . 'cart_rule_product_rule_value (id_product_rule, id_item)
+                    VALUES ' . implode(',', $productRuleValues)
+                )->executeStatement();
+            }
+        }
+    }
+
+    /**
+     * @param CartRuleId $cartRuleId
+     *
+     * @return RestrictionRuleGroup[]
+     */
+    public function getProductRestrictions(CartRuleId $cartRuleId): array
+    {
+        // retrieve all item ids based on cart rule
+        $ruleValues = $this->connection->createQueryBuilder()
+            ->select('crprv.id_product_rule, crprv.id_item')
+            ->from($this->dbPrefix . 'cart_rule_product_rule_value', 'crprv')
+            ->innerJoin(
+                'crprv',
+                $this->dbPrefix . 'cart_rule_product_rule',
+                'crpr',
+                'crprv.id_product_rule = crpr.id_product_rule'
+            )
+            ->innerJoin(
+                'crpr',
+                $this->dbPrefix . 'cart_rule_product_rule_group',
+                'crprg',
+                'crpr.id_product_rule_group = crprg.id_product_rule_group'
+            )
+            ->where('crprg.id_cart_rule = :cartRuleId')
+            ->setParameter('cartRuleId', $cartRuleId->getValue())
+            ->execute()
+            ->fetchAllAssociative()
+        ;
+
+        // retrieve all rules based on cart rule
+        $rules = $this->connection->createQueryBuilder()
+            ->select('crpr.id_product_rule, crpr.id_product_rule_group, crpr.type')
+            ->from($this->dbPrefix . 'cart_rule_product_rule', 'crpr')
+            ->innerJoin(
+                'crpr',
+                $this->dbPrefix . 'cart_rule_product_rule_group',
+                'crprg',
+                'crpr.id_product_rule_group = crprg.id_product_rule_group'
+            )
+            ->where('crprg.id_cart_rule = :cartRuleId')
+            ->setParameter('cartRuleId', $cartRuleId->getValue())
+            ->execute()
+            ->fetchAllAssociative()
+        ;
+
+        // retrieve all rule groups based on cart rule
+        $groups = $this->connection->createQueryBuilder()
```

Comment:
SAme question as the comment before, maybe even these three queries can be merged no?